text/nodejs/0013-install-bom.md

```diff
@@ -0,0 +1,161 @@
+
+For the directly installed dependencies in the Node Engine and Yarn buildpacks,
+we control the metadata around the dependencies and largely implemented BOM
+metadata retrieval on the dependency server side. The availability of this
+metadata made it straightforward to consume the metadata and generate the BOM.
+
+For the indirectly installed modules that are installed by the NPM Install and
+Yarn Install buildpacks, this is not the case because we delegate the
+installation process to the package managers. The case in which we have no
+package manager and use vendored node modules also falls under this case. Due
+to the nature of these dependencies and how they are installed, the process of
+generating a Bill of Materials will require changes to the language family itself.
+
+This is an important piece of the Bill of Materials, because it gives a
+complete picture of what dependencies are inside of the final application
+image. Without this metadata, a major value-add of a Bill of Materials is lost
+and users may have to use third-party tools to generate the rest of the BOM,
+which is a less than ideal user experience.
+
+
+## Implementation
+
+### Tool details
+
+The Bill of Materials that will result should be an accurate accounting of all
+of the modules that get installed into the `node_modules` directory.
+
+There already exists a tool called [CycloneDX Node.js
+Module](https://github.com/CycloneDX/cyclonedx-node-module) which is an
+official CycloneDX supported-tool that creates valid CycloneDX BOMs for Node.JS
+applications. CycloneDX is a BOM format supported by the OWASP Foundation, and
+is a serious contender (at the time of this RFC creation) as a format we will
+eventually support.
+
+The CycloneDX Node.js Module tool can easily be installed via `npm install` and
+then run from a directory containing `node_modules`. The output BOM contains
+the following fields for each module:
+* name,
+* version,
+* description,
+* hash,
+* source URI,
+* license ID,
+* and package URL
+
+All of these are fields that we aim to support per [the Paketo BOM
+RFC](https://github.com/sophiewigmore/rfcs/blob/8b1e8c9ed6201313f47c5897223cbffb265e96ed/text/0028-bill-of-materials.md)
+
+In order to support this tool in both the online and offline cases, we will
+pre-compile the tool with it's dependencies, and provide it as a dependency
+hosted on the [dep-server](https://github.com/paketo-buildpacks/dep-server).
+
+### Language Family Additions
+
+Using the CycloneDX tool to generate the BOM will involve some changes to the
+Node.JS language family buildpack order groups.
+
+* A single BOM generation buildpack will be added. It's primary functions are:
+  * perform the BOM dependency tool retrieval from the dep-server,
+  * execute BOM generation with the tool (`cyclonedx-bom -o bom.json`), and
+  * contribute the BOM to the `launch.toml` `[bom]` label.
+
+* For each of the three Node.JS buildpack order groups, the BOM generation
```

Comment:
I wonder if instead of generating the BOM during the build phase of the BOM generation buildpack, we should instead have the buildpack that produces installs the modules generate the BOM. There are a couple reasons for this:

Once CNB introduces the ability [associate BOM entries with a specific layer](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md), we will want to:
1. e.g. associate these BOM entries with a layer like paketo-buildpacks/yarn-install:modules
2. save time by avoiding rerunning the tool if the layer has not changed (the yarn-install buildpack should be responsible for making this determination)

In the case where there is no package manage we will still need the new buildpack to generate the BOM, or I guess node-start could potentially do it?

Reply:
I think that we need a new buildpack regardless and as it stands we can get away with having one buildpack do all of the BOM generation. I would prefer to implement the smallest change first and then potentially expand the functionality of other buildpacks or add additional buildpacks as the need arises. I have some concerns about having to duplicate logic across three different buildpacks just so I have the BOM on the right layer and so I would like to dive into this deeper when the feature is ready to play.

Is the lack of consideration for layer specific BOM a blocker for this initial implementation in your eyes?

Reply:
We just added a performance section, but when we checked the time it took to run the tool against an app with 1300 node modules, it only took around 3 seconds. It's pretty quick, so we're not overly worried about the re-running of the tool

Reply:
> it only took around 3 seconds. It's pretty quick, so we're not overly worried about the re-running of the tool

While I don't think 3 seconds is a deal-breaker, it's not completely insignificant either. For example a no-op rebuild of the paketo yarn sample app currently takes about 4.6 seconds on my workstation.

>  I have some concerns about having to duplicate logic across three different buildpacks just so I have the BOM on the right layer and so I would like to dive into this deeper when the feature is ready to play.

My assumption would be that the vast majority of the logic lives in the bom generation tool itself, leaving very little logic that would need to be duplicated?

> Is the lack of consideration for layer specific BOM a blocker for this initial implementation in your eyes?

No. In general I don't see my suggestions here as a blocker. I think associating the BOM with the correct layer and shaving a few seconds off the build whenever possible are both good long term goals, but if the node maintainers believe that accomplishing those goals would significantly slow down our ability to deliver ther feature or make the buildpacks harder to maintain, I trust the maintainers weigh those tradeoffs. This also seems like something that could be refactored later if need be without impacting end users, which lowers the stakes.

Reply:
> While I don't think 3 seconds is a deal-breaker, it's not completely insignificant either. For example a no-op rebuild of the paketo yarn sample app currently takes about 4.6 seconds on my workstation.

This is fair. The app we used is quite large so I'd expect that to be more of a worst-case time. Regardless, you are right though that it's significant enough that we should probably consider optimizing.
 
> My assumption would be that the vast majority of the logic lives in the bom generation tool itself, leaving very little logic that would need to be duplicated?

Right now the logic includes the installation of the tool (/cache reuse of the tool), running the tool, and converting the BOM JSON file into `packit.BOMEntry` types (for now). 

It's not terrible to duplicate, but the fact of the matter is that we can get the basic functionality for the BOM via a single buildpack, rather than adding a separate buildpack for one case, and making changes to NPM Install and Yarn Install in the other cases. The optimizations for build time and layer inclusion are definitely worthwhile, but I am more inclined to go with the basic implementation that we have outlined first (pending maintainer approval) and then work towards these optimizations down the line as you said.

@paketo-buildpacks/nodejs-maintainers can you weigh in?

Reply:
SGTM for the first implementation. If the `node_modules` hasn't changed, you should still be able to save time by avoiding downloading/rerunning the tool, right?

Reply:
Correct

Reply:
I'm also on board with the single-buildpack proposal as a first pass